src/v/archival/archiver_operations_impl.cc

```diff
@@ -299,6 +447,259 @@ class archiver_operations_impl : public archiver_operations_api {
+        }
+        vassert(false, "unreachable");
+    }
+
+    static bool segment_meta_matches_stats(
+      const cloud_storage::segment_meta& meta,
+      const cloud_storage::segment_record_stats& stats,
+      retry_chain_logger& ctxlog) {
+        // Validate segment content. The 'stats' is computed when
+        // the actual segment is scanned and represents the 'ground truth' about
+        // its content. The 'meta' is the expected segment metadata. We
+        // shouldn't replicate it if it doesn't match the 'stats'.
+        if (
+          meta.size_bytes != stats.size_bytes
+          || meta.base_offset != stats.base_rp_offset
+          || meta.committed_offset != stats.last_rp_offset
+          || static_cast<size_t>(meta.delta_offset_end - meta.delta_offset)
+               != stats.total_conf_records) {
+            vlog(
+              ctxlog.error,
+              "Metadata of the uploaded segment [size: {}, base: {}, last: {}, "
+              "begin delta: {}, end delta: {}] "
+              "doesn't match the segment [size: {}, base: {}, last: {}, total "
+              "config records: {}]",
+              meta.size_bytes,
+              meta.base_offset,
+              meta.committed_offset,
+              meta.delta_offset,
+              meta.delta_offset_end,
+              stats.size_bytes,
+              stats.base_rp_offset,
+              stats.last_rp_offset,
+              stats.total_conf_records);
+            return false;
+        }
+        return true;
+    }
+
+    std::tuple<remote_segment_path, ss::sstring, remote_segment_path>
+    segment_index_object_names(
+      const ss::shared_ptr<cluster_partition_api>& part,
+      const reconciled_upload_candidate_ptr& upl) {
+        auto path = part->get_remote_segment_path(upl->metadata);
+        auto ix = fmt::format("{}.index", path().native());
+        auto tx = fmt::format("{}.tx", path().native());
+        return std::make_tuple(path, ix, remote_segment_path(tx));
+    }
+
+    /// Cloning upload with the exception of tx data
+    std::tuple<ss::input_stream<char>, ss::input_stream<char>>
+    clone_stream(ss::input_stream<char> source) {
+        auto res = input_stream_fanout<2>(std::move(source), _readahead());
+
+        auto [lhs, rhs] = std::move(res);
+        return std::make_tuple(std::move(lhs), std::move(rhs));
+    }
+
+    /// Upload segment with index and tx-manifest
+    ///
+    /// The result is a combination of all upload results
+    /// plus the segment metadata generated during the index build
+    /// step.
+    ss::future<aggregated_upload_result> upload_candidate(
+      std::reference_wrapper<retry_chain_node> workflow_rtc,
+      ss::shared_ptr<cluster_partition_api> partition,
+      reconciled_upload_candidate_ptr upl) {
+        auto [segment_name, index_name, tx_name] = segment_index_object_names(
+          partition, upl);
+
+        cloud_storage::segment_record_stats stats{};
```

Comment:
I see this is used by upload_segment_index() but isn't used after. Should this be constructed and owned by upload_segment_index()? Or should this be included as a part of the returned result?